Support import.meta.main

It seems we would transform `import.meta.main` to `(require.main === module)` https://stackoverflow.com/a/31643112/188246

```
npm/src/src/build.ts:193:17 - error TS2339: Property 'main' does not exist on type 'ImportMeta'.

193 if (import.meta.main) {
```

--- a/lib/compiler_transforms.test.ts
+++ b/lib/compiler_transforms.test.ts
@@ -31,6 +31,6 @@ Deno.test("transform import.meta.main expressions", () => {
     "if (import.meta.main) { console.log('main'); }",
     `if ((require.main === module)) {
     console.log("main");
-}\n`);
+}\n`,
+  );
 });
-

--- a/lib/compiler_transforms.test.ts
+++ b/lib/compiler_transforms.test.ts
@@ -4,10 +4,10 @@ import { assertEquals } from "./test.deps.ts";
 import { ts } from "./mod.deps.ts";
 import { transformImportMeta } from "./compiler_transforms.ts";
 
-Deno.test("transform import.meta.url expressions", () => {
+function testImportReplacements(input: string, output: string) {
   const sourceFile = ts.createSourceFile(
     "file.ts",
-    "function test() { new URL(import.meta.url); }",
+    input,
     ts.ScriptTarget.Latest,
   );
   const newSourceFile =
@@ -16,8 +16,21 @@ Deno.test("transform import.meta.url expressions", () => {
     newLine: ts.NewLineKind.LineFeed,
   }).printFile(newSourceFile);
 
-  assertEquals(
-    text,
+  assertEquals(text, output);
+}
+
+Deno.test("transform import.meta.url expressions", () => {
+  testImportReplacements(
+    "function test() { new URL(import.meta.url); }",
     `function test() { new URL(require("url").pathToFileURL(__filename).href); }\n`,
   );
 });
+
+Deno.test("transform import.meta.main expressions", () => {
+  testImportReplacements(
+    "if (import.meta.main) { console.log('main'); }",
+    `if ((require.main === module)) {
+    console.log("main");
+}\n`);
+});
+

--- a/lib/compiler_transforms.ts
+++ b/lib/compiler_transforms.ts
@@ -6,6 +6,8 @@ import { ts } from "./mod.deps.ts";
 export const transformImportMeta: ts.TransformerFactory<ts.SourceFile> = (
   context,
 ) => {
+  const factory = context.factory;
+
   return (sourceFile) => ts.visitEachChild(sourceFile, visitNode, context);
 
   function visitNode(node: ts.Node): ts.Node {
@@ -14,16 +16,19 @@ export const transformImportMeta: ts.TransformerFactory<ts.SourceFile> = (
       ts.isPropertyAccessExpression(node) &&
       ts.isMetaProperty(node.expression) &&
       node.expression.keywordToken === ts.SyntaxKind.ImportKeyword &&
-      ts.isIdentifier(node.name) &&
-      node.name.escapedText === "url"
+      ts.isIdentifier(node.name)
     ) {
-      return getReplacementBinaryExpr();
+      if (node.name.escapedText === "url") {
+        return getReplacementImportMetaUrl();
+      } else if (node.name.escapedText === "main") {
+        return getReplacementImportMetaMain();
+      }
     }
+
     return ts.visitEachChild(node, visitNode, context);
   }
 
-  function getReplacementBinaryExpr(): ts.PropertyAccessExpression {
-    const factory = context.factory;
+  function getReplacementImportMetaUrl() {
     // Copy and pasted from ts-ast-viewer.com
     // require("url").pathToFileURL(__filename).href
     return factory.createPropertyAccessExpression(
@@ -41,5 +46,18 @@ export const transformImportMeta: ts.TransformerFactory<ts.SourceFile> = (
       ),
       factory.createIdentifier("href"),
     );
+  }
+
+  function getReplacementImportMetaMain() {
+    // Copy and pasted from ts-ast-viewer.com
+    // (require.main === module)
+    return factory.createParenthesizedExpression(factory.createBinaryExpression(
+      factory.createPropertyAccessExpression(
+        factory.createIdentifier("require"),
+        factory.createIdentifier("main")
+      ),
+      factory.createToken(ts.SyntaxKind.EqualsEqualsEqualsToken),
+      factory.createIdentifier("module")
+    ))
   }
 };

--- a/lib/compiler_transforms.ts
+++ b/lib/compiler_transforms.ts
@@ -54,10 +54,10 @@ export const transformImportMeta: ts.TransformerFactory<ts.SourceFile> = (
     return factory.createParenthesizedExpression(factory.createBinaryExpression(
       factory.createPropertyAccessExpression(
         factory.createIdentifier("require"),
-        factory.createIdentifier("main")
+        factory.createIdentifier("main"),
       ),
       factory.createToken(ts.SyntaxKind.EqualsEqualsEqualsToken),
-      factory.createIdentifier("module")
-    ))
+      factory.createIdentifier("module"),
+    ));
   }
 };

